File: 0016-3sum-closest/0016-3sum-closest.cpp
#include <algorithm>
#include <vector>
#include <limits>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        int closest_sum = nums[0] + nums[1] + nums[2];
        int min_diff = abs(target - closest_sum);
        
        for (int i = 0; i < n - 2; i++) {
            int left = i + 1;
            int right = n - 1;
            
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                int diff = abs(target - sum);
                
                if (diff < min_diff) {
                    min_diff = diff;
                    closest_sum = sum;
                }
                
                if (sum < target) {
                    left++;
                } else if (sum > target) {
                    right--;
                } else {
                    return target; // The sum is equal to the target, so it's the closest.
                }
            }
        }
        
        return closest_sum;
    }
};

File: 0016-3sum-closest/0016-3sum-closest.swift
class Solution {
    func threeSumClosest(_ nums: [Int], _ target: Int) -> Int {
        let n = nums.count
        let sortedNums = nums.sorted()
        var closestSum = sortedNums[0] + sortedNums[1] + sortedNums[2]
        var minDiff = abs(target - closestSum)

        for i in 0..<n - 2 {
            var left = i + 1
            var right = n - 1

            while left < right {
                let sum = sortedNums[i] + sortedNums[left] + sortedNums[right]
                let diff = abs(target - sum)

                if diff < minDiff {
                    minDiff = diff
                    closestSum = sum
                }

                if sum < target {
                    left += 1
                } else if sum > target {
                    right -= 1
                } else {
                    return target // The sum is equal to the target, so it's the closest.
                }
            }
        }

        return closestSum
    }
}
